### PR TITLE
Azure/redis cache schedule updates

### PR DESCRIPTION
**Pull request by alphadev4**

### CODE REVIEW

File: plugins/azure/redisCache/redisCacheDiagnosticLogs.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+module.exports = {
+    title: 'Redis Cache Diagnostic Logs Enabled',
+    category: 'Redis Cache',
+    domain: 'Databases',
+    description: 'Ensures diagnostic logging is enabled for Azure Cache for Redis.',
+    more_info: 'Enabling diagnostic setting helps you understand who is connecting to your caches and the timestamp of those connections. The log data could be used to identify the scope of a security breach and for security auditing purposes.',
+    recommended_action: 'Enable diagnostic logging for all Redis Caches.',
+    link: 'https://learn.microsoft.com/en-us/azure/azure-cache-for-redis/cache-monitor-diagnostic-settings?tabs=basic-standard-premium',
+    apis: ['redisCaches:listBySubscription','diagnosticSettings:listByRedisCache'],
```
Comment (mehakseedat63) [inline plugins/azure/redisCache/redisCacheDiagnosticLogs.js:12]:
```suggestion
    apis: ['redisCaches:listBySubscription','diagnosticSettings:listByRedisCache'],
```
why did we check this out from the other branch when they're using different calls?

File: plugins/azure/redisCache/redisCacheDiagnosticLogs.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                const diagnosticSettings = helpers.addSource(cache, source,
+                    ['diagnosticSettings', 'listByRedisCache', location, redisCache.id]);
+
+                if (!diagnosticSettings || diagnosticSettings.err || !diagnosticSettings.data) {
+                    helpers.addResult(results, 3, 'Unable to query Redis Cache diagnostics settings: ' + helpers.addError(diagnosticSettings), location, redisCache.id);
+                } else {
+                    var redisCacheDiagnosticLogs = false;
+                    diagnosticSettings.data.forEach(setting => {
+                        var logs = setting.logs;
+                        if (logs.some(log => (log.categoryGroup === 'audit' || log.categoryGroup === 'allLogs' || log.category === 'ConnectedClientList') && log.enabled)) {
```
Comment (mehakseedat63) [inline plugins/azure/redisCache/redisCacheDiagnosticLogs.js:47]:
```suggestion
                        if (logs.some(log => (log.categoryGroup === 'audit' || log.categoryGroup === 'allLogs' || log.category === 'ConnectedClientList') && log.enabled)) {
```
can we make this a setting to allow users to select which logs should be enabled?

File: plugins/azure/redisCache/redisCacheDiagnosticLogs.spec.js [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        'location': 'East US',
+        'name': 'test-cache',
+        'type': 'Microsoft.Cache/Redis',
+        'minimumTlsVersion': '1.1',
+    },
+];
+
+const diagnosticSettings = [
+    {
+        id: '/subscriptions/26a1a07e-06dd-4892-92c9-e4996b0fc546/resourcegroups/cloudsploit-dev/providers/microsoft.cache/redis/omerredistest/providers/microsoft.insights/diagnosticSettings/test',
```
Comment (mehakseedat63) [inline plugins/azure/redisCache/redisCacheDiagnosticLogs.spec.js:23]:
```suggestion
        id: '/subscriptions/26a1a07e-06dd-4892-92c9-e4996b0fc546/resourcegroups/cloudsploit-dev/providers/microsoft.cache/redis/omerredistest/providers/microsoft.insights/diagnosticSettings/test',
```
change this

**Review by mehakseedat63 — CHANGES_REQUESTED** [review]

**Review by mehakseedat63 — CHANGES_REQUESTED** [review]

File: plugins/azure/redisCache/redisCacheDiagnosticLogs.js (hunk 2 above)
Comment (alphadev4) [reply plugins/azure/redisCache/redisCacheDiagnosticLogs.js:47]: the minimum requirement mentioned in the plugin description was to have these logs enabled.

File: plugins/azure/redisCache/redisCacheScheduledUpdates.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+                return rcb();
+            }
+
+            caches.data.forEach(function(redisCache) {
+                if (!redisCache.id) return;
+                const patchSchedules = helpers.addSource(cache, source,
+                    ['patchSchedules', 'listByRedisCache', location, redisCache.id]);
+
+                if (!patchSchedules || (patchSchedules && patchSchedules.err)) {
+                    if (patchSchedules.err && patchSchedules.err.includes('There are no patch schedules found for redis cache')) {
```
[the remark this answers is not in the document]
Comment (alphadev4) [reply plugins/azure/redisCache/redisCacheScheduledUpdates.js:41]: Error code is same for two calls, if there are no patches we get the Error code as ResourceNotFound, and this code is same if call is made for wrong resource. So i think its better to check for the specific message that we get

File: plugins/azure/redisCache/redisCacheDiagnosticLogs.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                        diagnosticSettings.data.forEach(settings => {
+                            const logs = settings.logs;
+                            missingLogs = missingLogs.filter(requiredCategory =>
+                                !logs.some(log => (log.category && log.category.toLowerCase() === requiredCategory && log.enabled) || log.categoryGroup === 'allLogs' && log.enabled)
+                            );
+                        });
+
+                    }
+                    if (!missingLogs.length && found) {
+                        helpers.addResult(results, 0, 'Redis Cache has diagnostic logs enabled.', location, redisCache.id);
```
Comment (alphadev4) [inline plugins/azure/redisCache/redisCacheDiagnosticLogs.js:72]:
```suggestion
                        helpers.addResult(results, 0, 'Redis Cache has diagnostic logs enabled', location, redisCache.id);
```

**Review by mehakseedat63 — APPROVED** [review]

File: plugins/azure/redisCache/redisCacheDiagnosticLogs.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                            missingLogs = missingLogs.filter(requiredCategory =>
+                                !logs.some(log => (log.category && log.category.toLowerCase() === requiredCategory && log.enabled) || log.categoryGroup === 'allLogs' && log.enabled)
+                            );
+                        });
+
+                    }
+                    if (!missingLogs.length && found) {
+                        helpers.addResult(results, 0, 'Redis Cache has diagnostic logs enabled', location, redisCache.id);
+                    } else {
+                        helpers.addResult(results, 2, `Redis Cache does not have diagnostic logs enabled ${missingLogs.length? `for following: ${missingLogs}`: ''}`, location, redisCache.id);
```
Comment (alphadev4) [inline plugins/azure/redisCache/redisCacheDiagnosticLogs.js:74]:
```suggestion
                        helpers.addResult(results, 2, `Redis Cache does not have diagnostic logs enabled${missingLogs.length? ` for following: ${missingLogs}`: ''}`, location, redisCache.id);
```

**Review by mehakseedat63 — APPROVED** [review]

File: plugins/azure/redisCache/redisCacheDiagnosticLogs.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                            missingLogs = missingLogs.filter(requiredCategory =>
+                                !logs.some(log => (log.category && log.category.toLowerCase() === requiredCategory && log.enabled) || log.categoryGroup === 'allLogs' && log.enabled)
+                            );
+                        });
+
+                    }
+                    if (!missingLogs.length && found) {
+                        helpers.addResult(results, 0, 'Redis Cache has diagnostic logs enabled', location, redisCache.id);
+                    } else {
+                        helpers.addResult(results, 2, `Redis Cache does not have diagnostic logs enabled${missingLogs.length? ` for following: ${missingLogs}`: ''}`, location, redisCache.id);
```
Comment (alphadev4) [inline plugins/azure/redisCache/redisCacheDiagnosticLogs.js:74]:
```suggestion
                        helpers.addResult(results, 2, `Redis Cache does not have diagnostic logs enabled ${missingLogs.length? `for following: ${missingLogs}`: ''}`, location, redisCache.id);
```

**Pull request merged**